rllib/env/single_agent_episode.py

```diff
@@ -330,42 +330,35 @@ def __init__(
                     data=v, lookback=len_lookback_buffer
                 )
 
-        # TODO (sven): Remove this in favor of logging render images from an env inside
-        #  custom callbacks and using a to-be-designed metrics logger. Render images
-        #  from the env should NOT be stored in an episode (b/c they have nothing to do
-        #  with the data to be learned from, which should be the only thing an episode
-        #  has to be concerned with).
-        # RGB uint8 images from rendering the env; the images include the corresponding
-        # rewards.
-        assert render_images is None or observations is not None
-        self.render_images = render_images or []
-
         # The (global) timestep when this episode (possibly an episode chunk) started,
         # excluding a possible lookback buffer.
         self.t_started = t_started or 0
         # The current (global) timestep in the episode (possibly an episode chunk).
         self.t = len(self.rewards) + self.t_started
 
+        # Caches for temporary per-timestep data. May be used to store custom metrics
+        # from within a callback for the ongoing episode (e.g. render images).
+        self._temporary_timestep_data = defaultdict(list)
```

Comment:
Sweet. But how can `gym.Env`s use it? For example, if I have custom metrics coming from my env - I have to put it into the `info`.

Reply:
Ah I get it. Then the follow-up question is: Can I rewrite the episode while ongoing?

Reply:
The `_temporary_timestep_data` is really NOT affecting the episode in any way. Episode here merely serves as a "dumb" container for temporarily storing per-timestep custom stuff.

But yes, you can alter the episode via the setter APIs. Some of our ConnectorV2s take advantage of that, but it's not fully hashed out yet (still very alpha'ish).